### PR TITLE
Use one thread to refresh TopicMetadata and CConfiguration

### PR DESCRIPTION
**Pull request by gokulavasan**
Instead of using one each per region.

JIRA: https://issues.cask.co/browse/CDAP-8324

Build: http://builds.cask.co/browse/CDAP-RUT551

### CODE REVIEW

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java [hunk 1]
```diff
@@ -54,20 +54,29 @@ public CConfiguration getCConf() {
...
+      refreshThread.join(TimeUnit.SECONDS.toMillis(1));
     }
   }
 
   private void startRefreshThread() {
     refreshThread = new Thread("cdap-configuration-cache-refresh") {
       @Override
       public void run() {
-        while (!isInterrupted()) {
+        while ((!isInterrupted()) && isRunning()) {
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCache.java]: `isRunning` will return false before `nofityStarted()` has been called for `AbstractIdleService`

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCacheSupplier.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * Supplies instances of {@link CConfigurationCache} implementations.
+ */
+public class CConfigurationCacheSupplier implements CacheSupplier<CConfigurationCache> {
+  private static final Log LOG = LogFactory.getLog(CConfigurationCacheSupplier.class);
+
+  private static volatile CConfigurationCache instance;
+  private static volatile int refCount = 0;
+  private static volatile Object lock = new Object();
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCacheSupplier.java]: `lock` should be final

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCacheSupplier.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+package co.cask.cdap.data2.transaction.coprocessor;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
+
+/**
+ * Supplies instances of {@link CConfigurationCache} implementations.
+ */
+public class CConfigurationCacheSupplier implements CacheSupplier<CConfigurationCache> {
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCacheSupplier.java:26]: We can turn this into an abstract class with a protected method `Service createInstance()`, right?

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CacheSupplier.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+import com.google.common.base.Supplier;
+import com.google.common.util.concurrent.AbstractIdleService;
+
+/**
+ * Provides ability to get and release obejcts
+ *
+ * @param <T> type of the object supplied
+ */
+public interface CacheSupplier<T extends AbstractIdleService> extends Supplier<T> {
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CacheSupplier.java]: We can use `T extends Service` here

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CacheSupplier.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+  /**
+   * @return Get an instance of T
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CacheSupplier.java]: It would be good to add that if this is the first object then it will be started, and similar comment for release

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/messaging/TopicMetadataCache.java [hunk 6]
```diff
@@ -83,9 +84,18 @@ public boolean isAlive() {
     return refreshThread.isAlive();
   }
 
-  public void stop() {
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting TopicMetadataCache Refresh Thread.");
+    startRefreshThread();
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/messaging/TopicMetadataCache.java:90]: I see that the refresh thread is also being started in the constructor

**Review by poornachandra — COMMENTED** [review]
@gokulavasan I have a few comments

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCacheSupplier.java (hunk 3 above)
Comment (gokulavasan) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCacheSupplier.java:26]: Is this so that we can derive from this abstract class for the two suppliers? Any ideas on how we can handle the static variables - instance, object, refCount?

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/messaging/TopicMetadataCache.java (hunk 6 above)
Comment (gokulavasan) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/messaging/TopicMetadataCache.java:90]: Fixed.

**Comment by gokulavasan** [comment]
@poornachandra Addressed comments. Please take a look again

**Comment by gokulavasan** [comment]
@poornachandra Pulled out the common pieces of Reference counting into a separate class. Please review when you get a chance.

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCacheSupplier.java (hunk 3 above)
Comment (gokulavasan) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCacheSupplier.java:26]: Addressed this issue in the last commit.

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    synchronized (lock) {
+      int newCount = refCount.decrementAndGet();
+      if (LOG.isDebugEnabled()) {
+        LOG.debug(String.format("Decrementing reference count for %s: %d", instanceName, newCount));
+      }
+
+      if (newCount == 0) {
+        try {
+          Service serviceInstance = instance.get();
+          serviceInstance.stopAndWait();
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java]: Need to set `instance` to null after this

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/messaging/TopicMetadataCacheSupplier.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * Supplies instances of {@link TopicMetadataCache} implementations.
+ */
+public class TopicMetadataCacheSupplier implements CacheSupplier<TopicMetadataCache> {
+  private static AtomicReference<TopicMetadataCache> instance = new AtomicReference<>();
+  private static AtomicInteger refCount = new AtomicInteger(0);
+  private static final Object lock = new Object();
+
+  private final ReferenceCountedSupplier<TopicMetadataCache> referenceCountedSupplier;
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/messaging/TopicMetadataCacheSupplier.java]: If `referenceCountedSupplier` instance is made static then we can move `instance`, `lock` and refCount` as instance variable of class `ReferenceCountedSupplier`, right?

**Review by poornachandra — COMMENTED** [review]
A couple more comments @gokulavasan

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java (hunk 7 above)
Comment (poornachandra) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java]: Also it would be good to add a test where we bring down instances of the supplier to zero and then increase them again and then down to zero

Comment (poornachandra) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java]: It would be good to instantiate two different suppliers at the same time in the test to check multiple supplier types work correctly

Comment (gokulavasan) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java]: Fixed and enhanced tests. Thanks for the catch!

**Comment by gokulavasan** [comment]
@poornachandra Addressed comments in the last two commits. Please take a look.

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCacheSupplier.java [hunk 9]
```diff
@@ -20,37 +20,33 @@
...
 /**
  * Supplies instances of {@link CConfigurationCache} implementations.
  */
 public class CConfigurationCacheSupplier implements CacheSupplier<CConfigurationCache> {
-  private static final Object lock = new Object();
-  private static AtomicReference<CConfigurationCache> instance = new AtomicReference<>();
-  private static AtomicInteger refCount = new AtomicInteger(0);
-
-  private final ReferenceCountedSupplier<CConfigurationCache> referenceCountedSupplier;
+  private static volatile AtomicReference<ReferenceCountedSupplier<CConfigurationCache>> referenceCountedSupplier
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/transaction/coprocessor/CConfigurationCacheSupplier.java]: This can be changed to -
```
private static final ReferenceCountedSupplier<CConfigurationCache> = 
             new ReferenceCountedSupplier<>(CConfigurationCache.class.getSimpleName(), 
                                           new Supplier<CConfigurationCache>() {
                                              @Override
                                              public CConfigurationCache get() {
                                                 return new CConfigurationCache(hConf, sysConfTablePrefix);
                                             }
                                     });
```

This way we can ensure that initialization happens only once, and don't need the `AtomicReference`.

**Review by poornachandra — COMMENTED** [review]
@gokulavasan one comment. I'm yet to take a look at the tests.

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java [hunk 10]
```diff
@@ -31,23 +31,21 @@
...
-    this.instanceSupplier = instanceSupplier;
   }
 
-  @Override
-  public T get() {
+  public T getOrCreate(Supplier<T> instanceSupplier) {
+    // Fix the pessimistic locking here
     synchronized (lock) {
       if (instance.get() == null) {
         // Instance has not been instantiated
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java]: Let's add a debug log message when we create the instance

Comment (poornachandra) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java]: Same while stopping the instance too

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java [hunk 11]
```diff
@@ -63,8 +61,8 @@ public T get() {
     }
   }
 
-  @Override
   public void release() {
+    // Fix the pessimistic locking here
     synchronized (lock) {
       int newCount = refCount.decrementAndGet();
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java]: It would be a good idea to guard against more releases than gets.

Comment (poornachandra) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java]: And also add a test case for it

**Review by poornachandra — COMMENTED** [review]
@gokulavasan Few minor comments. Can you also add some comments to the test to help understand it?

**Comment by gokulavasan** [comment]
@poornachandra Addressed comments. Please take a look again.

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java [hunk 12]
```diff
@@ -64,18 +67,29 @@ public T getOrCreate(Supplier<T> instanceSupplier) {
   public void release() {
     // Fix the pessimistic locking here
     synchronized (lock) {
+      if (refCount.get() <= 0) {
+        if (LOG.isDebugEnabled()) {
+          LOG.debug(String.format("Reference Count for Service %s is already zero.", instanceName));
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java]: This should be a warn, no?

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        try {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug(String.format("Reference Count for Service is 0. Stopping Service : %s", instanceName));
+          }
+
+          Service serviceInstance = instance.get();
+          serviceInstance.stopAndWait();
+          instance.set(null);
+        } catch (Exception ex) {
+          LOG.warn(String.format("Exception while trying to stop Service %s.", instanceName));
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java]: Would be good to log the stack trace here

File: cdap-hbase-compat-base/src/test/java/co/cask/cdap/data2/util/ReferenceCountedSupplierTests.java [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Tests {@link ReferenceCountedSupplier}s.
+ */
+public class ReferenceCountedSupplierTests {
```
Comment (poornachandra) [inline cdap-hbase-compat-base/src/test/java/co/cask/cdap/data2/util/ReferenceCountedSupplierTests.java:41]: I see this exception when I run this test -
```
Exception in thread "cdap-configuration-cache-refresh" java.lang.NullPointerException
	at org.apache.hadoop.hbase.client.HTable.backgroundFlushCommits(HTable.java:964)
	at org.apache.hadoop.hbase.client.HTable.flushCommits(HTable.java:1252)
	at org.apache.hadoop.hbase.client.HTable.close(HTable.java:1289)
	at co.cask.cdap.data2.util.hbase.ConfigurationTable.read(ConfigurationTable.java:152)
	at co.cask.cdap.data2.transaction.queue.hbase.coprocessor.CConfigurationReader.read(CConfigurationReader.java:39)
	at co.cask.cdap.data2.transaction.coprocessor.CConfigurationCache$1.run(CConfigurationCache.java:85)
Exception in thread "tms-topic-metadata-cache-refresh" java.lang.NullPointerException
	at co.cask.cdap.messaging.TopicMetadataCache.updateCache(TopicMetadataCache.java:123)
	at co.cask.cdap.messaging.TopicMetadataCache$2.run(TopicMetadataCache.java:183)
```

Is it because we use nulls as constructor arguments of `CConfigurationCacheSupplier` in the test?

**Review by poornachandra — COMMENTED** [review]
@gokulavasan Just a couple of minor comments

File: cdap-hbase-compat-base/src/test/java/co/cask/cdap/data2/util/ReferenceCountedSupplierTests.java (hunk 14 above)
Comment (gokulavasan) [reply cdap-hbase-compat-base/src/test/java/co/cask/cdap/data2/util/ReferenceCountedSupplierTests.java:41]: Yes

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java (hunk 12 above)
Comment (gokulavasan) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java]: Ok I will change the level to WARN

File: cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java (hunk 13 above)
Comment (gokulavasan) [reply cdap-hbase-compat-base/src/main/java/co/cask/cdap/data2/util/ReferenceCountedSupplier.java]: Sure, will make that change.

**Comment by gokulavasan** [comment]
@poornachandra Addressed comments. Please take a look

**Comment by poornachandra** [comment]
LGTM
